### DISCUSSION TITLE
Can't import crazyswarm.svc when the testcode run

### DISCUSSION THREAD

**Discussion by imrd1123** [USC-ACTLab/crazyswarm#495] [General]
Environment
    OS=Ubuntu 20.04 (BitcrazeVM)
    ROS=Noetic
    Python=3.8.10
-------------------------------
Hi. I'm in trouble with installing crazyswarm.
I'm glad if you help to solve my problem.

Situation:
I'm trying to install crazyswarm to my PC by reading a official document.
[https://crazyswarm.readthedocs.io/en/latest/installation.html]
When I execute the following command, the error emerged.
`python3 -m pytest`
It seems my computer can't find the crazyswarm.srv file.

Try:
・attempt to find crazyswarm.srv from my computer -> can't find
・search issues -> no same problem

Error Log:
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-4.6.9, py-1.8.1, pluggy-0.13.0
rootdir: /home/bitcraze/apps/crazyswarm/ros_ws/src/crazyswarm/scripts, inifile: pytest.ini
collected 31 items                                                             

test_collisionAvoidance.py ............                                  [ 38%]
test_highLevel.py .........                                              [ 67%]
test_lowLevel.py ......                                                  [ 87%]
test_ros.py F                                                            [ 90%]
test_simOnly.py .                                                        [ 93%]
test_videoOutput.py F                                                    [ 96%]
test_yamlString.py .                                                     [100%]

=================================== FAILURES ===================================
_______________________________ test_ros_import ________________________________

    @pytest.mark.ros
    def test_ros_import():
>       from pycrazyswarm.crazyflie import TimeHelper, CrazyflieServer

test_ros.py:6: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    import sys
    import yaml
    import rospy
    import numpy as np
    import time
    import tf_conversions
    from std_srvs.srv import Empty
    import std_msgs
>   from crazyswarm.srv import *
E   ModuleNotFoundError: No module named 'crazyswarm'

pycrazyswarm/crazyflie.py:12: ModuleNotFoundError
_______________________________ test_videoOutput _______________________________

tmp_path = PosixPath('/tmp/pytest-of-bitcraze/pytest-2/test_videoOutput0')

    @pytest.mark.skipif("TRAVIS" in os.environ or "CI" in os.environ,
                        reason="CI usually cannot create OpenGL context.")
    def test_videoOutput(tmp_path):
        # tmp_path is supplied by pytest - a directory where we can write that will
        # eventually be deleted.
        path = str(tmp_path / "crazyswarm_test_video.mp4")
>       subprocess.call(["python", __file__, path])

test_videoOutput.py:64: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.8/subprocess.py:340: in call
    with Popen(*popenargs, **kwargs) as p:
/usr/lib/python3.8/subprocess.py:858: in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <subprocess.Popen object at 0x7f2d43db77f0>
args = ['python', '/home/bitcraze/apps/crazyswarm/ros_ws/src/crazyswarm/scripts/test_videoOutput.py', '/tmp/pytest-of-bitcraze/pytest-2/test_videoOutput0/crazyswarm_test_video.mp4']
executable = b'python', preexec_fn = None, close_fds = True, pass_fds = ()
cwd = None, env = None, startupinfo = None, creationflags = 0, shell = False
p2cread = -1, p2cwrite = -1, c2pread = -1, c2pwrite = -1, errread = -1
errwrite = -1, restore_signals = True, start_new_session = False

    def _execute_child(self, args, executable, preexec_fn, close_fds,
                       pass_fds, cwd, env,
                       startupinfo, creationflags, shell,
                       p2cread, p2cwrite,
                       c2pread, c2pwrite,
                       errread, errwrite,
                       restore_signals, start_new_session):
        """Execute program (POSIX version)"""
    
        if isinstance(args, (str, bytes)):
            args = [args]
        elif isinstance(args, os.PathLike):
            if shell:
                raise TypeError('path-like args is not allowed when '
                                'shell is true')
            args = [args]
        else:
            args = list(args)
    
        if shell:
            # On Android the default shell is at '/system/bin/sh'.
            unix_shell = ('/system/bin/sh' if
                      hasattr(sys, 'getandroidapilevel') else '/bin/sh')
            args = [unix_shell, "-c"] + args
            if executable:
                args[0] = executable
    
        if executable is None:
            executable = args[0]
    
        sys.audit("subprocess.Popen", executable, args, cwd, env)
    
        if (_USE_POSIX_SPAWN
                and os.path.dirname(executable)
                and preexec_fn is None
                and not close_fds
                and not pass_fds
                and cwd is None
                and (p2cread == -1 or p2cread > 2)
                and (c2pwrite == -1 or c2pwrite > 2)
                and (errwrite == -1 or errwrite > 2)
                and not start_new_session):
            self._posix_spawn(args, executable, env, restore_signals,
                              p2cread, p2cwrite,
                              c2pread, c2pwrite,
                              errread, errwrite)
            return
    
        orig_executable = executable
    
        # For transferring possible exec failure from child to parent.
        # Data format: "exception name:hex errno:description"
        # Pickle is not used; it is complex and involves memory allocation.
        errpipe_read, errpipe_write = os.pipe()
        # errpipe_write must not be in the standard io 0, 1, or 2 fd range.
        low_fds_to_close = []
        while errpipe_write < 3:
            low_fds_to_close.append(errpipe_write)
            errpipe_write = os.dup(errpipe_write)
        for low_fd in low_fds_to_close:
            os.close(low_fd)
        try:
            try:
                # We must avoid complex work that could involve
                # malloc or free in the child process to avoid
                # potential deadlocks, thus we do all this here.
                # and pass it to fork_exec()
    
                if env is not None:
                    env_list = []
                    for k, v in env.items():
                        k = os.fsencode(k)
                        if b'=' in k:
                            raise ValueError("illegal environment variable name")
                        env_list.append(k + b'=' + os.fsencode(v))
                else:
                    env_list = None  # Use execv instead of execve.
                executable = os.fsencode(executable)
                if os.path.dirname(executable):
                    executable_list = (executable,)
                else:
                    # This matches the behavior of os._execvpe().
                    executable_list = tuple(
                        os.path.join(os.fsencode(dir), executable)
                        for dir in os.get_exec_path(env))
                fds_to_keep = set(pass_fds)
                fds_to_keep.add(errpipe_write)
                self.pid = _posixsubprocess.fork_exec(
                        args, executable_list,
                        close_fds, tuple(sorted(map(int, fds_to_keep))),
                        cwd, env_list,
                        p2cread, p2cwrite, c2pread, c2pwrite,
                        errread, errwrite,
                        errpipe_read, errpipe_write,
                        restore_signals, start_new_session, preexec_fn)
                self._child_created = True
            finally:
                # be sure the FD is closed no matter what
                os.close(errpipe_write)
    
            self._close_pipe_fds(p2cread, p2cwrite,
                                 c2pread, c2pwrite,
                                 errread, errwrite)
    
            # Wait for exec to fail or succeed; possibly raising an
            # exception (limited in size)
            errpipe_data = bytearray()
            while True:
                part = os.read(errpipe_read, 50000)
                errpipe_data += part
                if not part or len(errpipe_data) > 50000:
                    break
        finally:
            # be sure the FD is closed no matter what
            os.close(errpipe_read)
    
        if errpipe_data:
            try:
                pid, sts = os.waitpid(self.pid, 0)
                if pid == self.pid:
                    self._handle_exitstatus(sts)
                else:
                    self.returncode = sys.maxsize
            except ChildProcessError:
                pass
    
            try:
                exception_name, hex_errno, err_msg = (
                        errpipe_data.split(b':', 2))
                # The encoding here should match the encoding
                # written in by the subprocess implementations
                # like _posixsubprocess
                err_msg = err_msg.decode()
            except ValueError:
                exception_name = b'SubprocessError'
                hex_errno = b'0'
                err_msg = 'Bad exception data from child: {!r}'.format(
                              bytes(errpipe_data))
            child_exception_type = getattr(
                    builtins, exception_name.decode('ascii'),
                    SubprocessError)
            if issubclass(child_exception_type, OSError) and hex_errno:
                errno_num = int(hex_errno, 16)
                child_exec_never_called = (err_msg == "noexec")
                if child_exec_never_called:
                    err_msg = ""
                    # The error must be from chdir(cwd).
                    err_filename = cwd
                else:
                    err_filename = orig_executable
                if errno_num != 0:
                    err_msg = os.strerror(errno_num)
>               raise child_exception_type(errno_num, err_msg, err_filename)
E               FileNotFoundError: [Errno 2] No such file or directory: 'python'

/usr/lib/python3.8/subprocess.py:1704: FileNotFoundError
===================== 2 failed, 29 passed in 6.91 seconds ======================

**Comment by whoenig** [comment]
The first failure can be avoided by `source ../../../devel/setup.bash` before running the test. This needs to be fixed in the documentation. The second one is a bug - this test doesn't seem to be run in CI, so we didn't catch it before. Tracked in issue #497.

**Reply by imrd1123** [reply]
I'm sorry for the late reply.
I didn't expect to get a reply so quickly.

I read the comment and saw technical discussion #497
I tried to execute pytest after apply `source ../../../devel/setup.bash`.
Then I saw the test passed except for ffmpeg.
(I understand ffmpeg is not need if I don't use camera function.(from #497 discussion))

I can go to the next step.
Thank you for teaching me a solution of install.
